### PR TITLE
Added GD tests

### PR DESCRIPTION
**Pull request by radarhere**
This code is duplicated at https://github.com/python-pillow/Pillow/blob/master/PIL/ImageFile.py#L87. The only difference made by removing it here is that `filename` may be `None` instead of an empty string.

### CODE REVIEW

**Comment by wiredfool** [comment]
First of all, we have no tests at all on the `GdImageFile.py`. Nothing references it. Anywhere. 

Also, when looking at the comments in the file, there's this:
```
# NOTE: This format cannot be automatically recognized, so the
# class is not registered for use with Image.open().  To open a
# gd file, use the GdImageFile.open() function instead.

# THE GD FORMAT IS NOT DESIGNED FOR DATA INTERCHANGE.  This
# implementation is provided for convenience and demonstrational
# purposes only.
```

So, it looks like it's not designed to be used from within the ImageFile process. So, it doesn't look like duplicated code.

**Comment by radarhere** [comment]
The GdimageFile class inherits from ImageFile. So the parameters passed in are handled by ImageFile's init method. Yes?

**Comment by radarhere** [comment]
Okay, I have now added basic tests.

**Comment by hugovk** [comment]
Is there some documentation about the GD uncompressed format somewhere?

Where's the test GD file from, or how was it generated? Do we know it's a valid GD image?

How do we know the changes are correct?

Without the changes, the test file gives:
```python
>>> from PIL import GdImageFile
>>> im = GdImageFile.open("/tmp/Downloads/gd")
>>> im.size
(65535, 10)
im.show()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "PIL/Image.py", line 1783, in show
    _show(self, title=title, command=command)
  File "PIL/Image.py", line 2537, in _show
    _showxv(image, **options)
  File "PIL/Image.py", line 2542, in _showxv
    ImageShow.show(image, title, **options)
  File "PIL/ImageShow.py", line 51, in show
    if viewer.show(image, title=title, **options):
  File "PIL/ImageShow.py", line 75, in show
    return self.show_image(image, **options)
  File "PIL/ImageShow.py", line 94, in show_image
    return self.show_file(self.save_image(image), **options)
  File "PIL/ImageShow.py", line 90, in save_image
    return image._dump(format=self.get_format(image))
  File "PIL/Image.py", line 576, in _dump
    self.load()
  File "PIL/ImageFile.py", line 171, in load
    self.map, self.size, decoder_name, extents, offset, args
ValueError: buffer is not large enough
```

With the changes, we get:
```python
>>> from PIL import GdImageFile
>>> im = GdImageFile.open("/tmp/Downloads/gd")
>>> im.size
(10, 20)
>>> im.show()
>>> im
<PIL.GdImageFile.GdImageFile image mode=L size=10x20 at 0x108B3AD50>
```

The image shows as a black rectangle in Mac Preview, but how can we verify this is correct? (e.g. ImageMagick cannot identify it.)

gd

The `transparent`/`transparency` key isn't tested, so docs would also be useful for confirming this change.

**Comment by radarhere** [comment]
I generated the image using PHP's GD library.

    <?php
    $im = imagecreatetruecolor(10, 20);
    imagegd($im, 'gd');

I changed the key for consistency with other formats.

**Comment by hugovk** [comment]
Would it be possible to save a Hopper image in GD format with PHP?

Or make an image with something a bit more than plain black, maybe just a line or two? Even better if it can include transparency.

Then we can compare it to a corresponding image in another format, and at least assert some properties like size.

**Comment by radarhere** [comment]
Okay. I used the following code to create a hopper image -

    <?php
    $im = imagecreatefromjpeg('/Users/andrewmurray/Dropbox/docs/pillow/Pillow/Tests/images/hopper.jpg');
    imagegd($im, 'hoppergd');

The resulting image loaded by Pillow is clearly not correct.

**Comment by hugovk** [comment]
The GD image opened and saved as JPG by PHP:
out2

The GD image opened and saved as JPG by Pillow:
hoppergd

Let's add this to #2373.

**Comment by hugovk** [comment]
Flipping that -1 to 1 flips the picture:
```python
self.tile = [("raw", (0, 0)+self.size, 775, ("L", 0, -1))]
```
```python
self.tile = [("raw", (0, 0)+self.size, 775, ("L", 0, 1))]
```
hoppergd

Changing those `"P"`s to `"L"` as suggested by the `FIXME` gives:

hoppergd

More work is needed! Note this test image is a GD v2 file because `i16(s[0:2])` is [65535](https://github.com/php/php-src/blob/6053987bc27e8dede37f437193a5cad448f99bce/ext/gd/libgd/gd_gd.c#L233-L240). And let's call the file hopper.gd.

**Comment by radarhere** [comment]
Thanks for that. While I'm happy to go with whatever the consensus is, my concern with naming the test file hopper.gd is giving the false impression that .gd is a usable extension.

**Comment by hugovk** [comment]
.gd is mentioned several times in the PHP source above.

**Comment by radarhere** [comment]
Fair enough! I mistakenly inferred that it wasn't valid from the comment in GdImageFile that 'This format cannot be automatically recognized'.

**Comment by wiredfool** [comment]
@radarhere I think that comment means that we can't figure out the file type from the content, as the file extension really doesn't factor into the open loop. (It's only taken into account when you save, and then only if a format isn't specified)

And. Another format that doesn't work.

**Comment by radarhere** [comment]
Okay, I've added the extension now.

**Comment by wiredfool** [comment]
Here's what I can find for a spec: https://github.com/libgd/libgd/commit/5ee392d6e8c1695ceff5ada045d3f4374e51ecef#diff-909ccc1a14087aa650c846156ae32572

GD and GD2 are listed as both proprietary and obsolete. On the bright side, GD2 can be recognized from the first 4 bytes. 

I suspect that the problem is in the interpretation of the palette, as there's the basis of the image there. OTOH, the palette is supposed to be 256*32bit, and we're starting at offset 775, so there's something odd there.

File: PIL/GdImageFile.py [hunk 1]
```diff
@@ -53,12 +46,12 @@ def _open(self):
...
 
         self.mode = "L"  # FIXME: "P"
-        self.size = i16(s[0:2]), i16(s[2:4])
+        self.size = i16(s[2:4]), i16(s[4:6])
 
         # transparency index
         tindex = i16(s[5:7])
         if tindex < 256:
-            self.info["transparent"] = tindex
+            self.info["transparency"] = tindex
```
Comment (homm) [inline PIL/GdImageFile.py]: This line is not covered by tests

**Comment by radarhere** [comment]
I've renamed the PR.

Note that the current code is what I believe to be most theoretically correct, while also running without errors - XBGR is clearly not correct, but that was done in order to keep the rest of the code that resulted from my current understanding. While I may not at all be on the right path, if I am, this will provide a starting point for the next person.

**Comment by aclark4life** [comment]
Thanks

**Pull request merged**